adam_core/observers/state.py

```diff
@@ -0,0 +1,147 @@
+        - precession (IAU-1976)
+        - nutation (IAU-1980 with IERS corrections)
+        - polar motion
+    This frame is retrieved through SPICE.
+
+    Parameters
+    ----------
+    code : str
+        MPC observatory code for which to find the states.
+    observation_times : `~astropy.time.core.Time` (N)
+        Epochs for which to find the observatory locations.
+    frame : {'equatorial', 'ecliptic'}
+        Return observer state in the equatorial or ecliptic J2000 frames.
+    origin : OriginCodes
+        The NAIF ID of the origin.
+
+    Returns
+    -------
+    observer_states : `~adam_core.coordinates.cartesian.CartesianCoordinates`
+        The state vectors of the observer in the desired frame
+        and measured from the desired origin.
+    """
+    if code not in OBSERVATORY_CODES:
+        err = f"{code} is not a valid MPC observatory code."
+        raise ValueError(err)
+
+    if frame == "ecliptic":
+        frame_spice = "ECLIPJ2000"
+    elif frame == "equatorial":
+        frame_spice = "J2000"
+    else:
+        err = "frame should be one of {'equatorial', 'ecliptic'}"
+        raise ValueError(err)
+
+    # Make sure SPICE is ready to go
+    setup_SPICE()
+
+    # Get observatory geodetic information
+    geodetics = OBSERVATORY_GEODETICS.select("code", code).values[0]
+
+    if np.any(np.isnan(geodetics)):
+        err = (
+            f"{code} is missing information on Earth-based geodetic coordinates. The MPC Obs Code\n"
+            "file may be missing this information or the observer is a space-based observatory.\n"
+            "Space observatories are currently not supported.\n"
+        )
+        raise ValueError(err)
+
+    # Grab Earth state vector (this function handles duplicate times)
+    state = get_perturber_state(OriginCodes.EARTH, times, frame=frame, origin=origin)
+
+    # If the code is 500 (geocenter), we can just return the Earth state vector
+    if code == "500":
+        r_obs = state.r
+        v_obs = state.v
+
+    # If not then we need to add a topocentric correction
+    else:
+        # Get observer location on Earth
+        longitude = geodetics[0]
+        cos_phi = geodetics[1]
+        sin_phi = geodetics[2]
+        sin_longitude = np.sin(np.radians(longitude))
+        cos_longitude = np.cos(np.radians(longitude))
+
+        # Calculate pointing vector from geocenter to observatory
+        o_hat_ITRF93 = np.array(
+            [cos_longitude * cos_phi, sin_longitude * cos_phi, sin_phi]
+        )
+
+        # Multiply pointing vector with Earth radius to get actual vector
+        o_vec_ITRF93 = np.dot(R_EARTH, o_hat_ITRF93)
+
+        # Convert MJD epochs in TDB to ET in TDB
+        epochs_tdb = times.tdb.jd
+        unique_epochs_tdb = np.unique(epochs_tdb)
+        unique_epochs_et = np.array(
+            [sp.str2et("JD {:.16f} TDB".format(i)) for i in unique_epochs_tdb]
+        )
```

Comment:
Is this slow? This looks slow.

I don't think it ought to be necessary to call spicepy for this. I _think_ it should be possible to do this conversion yourself, since it's really just a change of epoch from JD epoch up to J2000. It might even be a constant shift for all values?

This is only worth investigating if it is indeed slow.

Reply:
Yes, it's definitely slow. The conversion to ephemeris time in `spiceypy` requires the leapseconds kernel so I've just assumed it's doing something a bit more complex than what I can do outside of `spiceypy`, but TDB to ET should be nearly the same... Let me change it and see if the unit test tolerance changes at all.